srml/staking/reward-curve/src/lib.rs

```diff
@@ -0,0 +1,406 @@
+		)));
+	}
+
+	Ok(value)
+}
+
+impl Parse for INposInput {
+	fn parse(input: ParseStream) -> syn::Result<Self> {
+		let args_input;
+
+		<syn::Token![const]>::parse(&input)?;
+		let ident = <syn::Ident>::parse(&input)?;
+		<syn::Token![:]>::parse(&input)?;
+		let typ = <syn::Type>::parse(&input)?;
+		<syn::Token![=]>::parse(&input)?;
+		<keyword::curve>::parse(&input)?;
+		<syn::Token![!]>::parse(&input)?;
+		syn::parenthesized!(args_input in input);
+		<syn::Token![;]>::parse(&input)?;
+
+		if !input.is_empty() {
+			return Err(input.error("expected end of input stream, no token expected"));
+		}
+
+		let min_inflation = parse_field::<keyword::min_inflation>(&args_input, Bounds {
+			min: 0,
+			min_strict: true,
+			max: 1_000_000,
+			max_strict: false,
+		})?;
+		<syn::Token![,]>::parse(&args_input)?;
+		let max_inflation = parse_field::<keyword::max_inflation>(&args_input, Bounds {
+			min: min_inflation,
+			min_strict: true,
+			max: 1_000_000,
+			max_strict: false,
+		})?;
+		<syn::Token![,]>::parse(&args_input)?;
+		let ideal_stake = parse_field::<keyword::ideal_stake>(&args_input, Bounds {
+			min: 0_100_000,
+			min_strict: false,
+			max: 0_900_000,
+			max_strict: false,
+		})?;
+		<syn::Token![,]>::parse(&args_input)?;
+		let falloff = parse_field::<keyword::falloff>(&args_input, Bounds {
+			min: 0_010_000,
+			min_strict: false,
+			max: 1_000_000,
+			max_strict: false,
+		})?;
+		<syn::Token![,]>::parse(&args_input)?;
+		let max_piece_count = parse_field::<keyword::max_piece_count>(&args_input, Bounds {
+			min: 2,
+			min_strict: false,
+			max: 1_000,
+			max_strict: false,
+		})?;
+		<syn::Token![,]>::parse(&args_input)?;
+		let test_precision = parse_field::<keyword::test_precision>(&args_input, Bounds {
+			min: 0,
+			min_strict: false,
+			max: 1_000_000,
+			max_strict: false,
+		})?;
+		<Option<syn::Token![,]>>::parse(&args_input)?;
+
+		if !args_input.is_empty() {
+			return Err(args_input.error("expected end of input stream, no token expected"));
+		}
+
+		Ok(Self {
+			ident,
+			typ,
+			min_inflation,
+			ideal_stake,
+			max_inflation,
+			falloff,
+			max_piece_count,
+			test_precision,
+		})
+	}
+}
+
+struct INPoS {
+	i_0: u32,
+	i_ideal_times_x_ideal: u32,
+	i_ideal: u32,
+	x_ideal: u32,
+	d: u32,
+}
+
+impl INPoS {
+	fn from_input(input: &INposInput) -> Self {
+		INPoS {
+			i_0: input.min_inflation,
+			i_ideal: (input.max_inflation as u64 * MILLION as u64 / input.ideal_stake as u64) as u32,
+			i_ideal_times_x_ideal: input.max_inflation,
+			x_ideal: input.ideal_stake,
+			d: input.falloff,
+		}
+	}
+
+	fn compute_opposite_after_x_ideal(&self, y: u32) -> u32 {
+		if y == self.i_0 {
+			return u32::max_value();
+		}
+		let log = log2(self.i_ideal_times_x_ideal - self.i_0, y - self.i_0);
+
+		self.x_ideal + ((self.d as u64 * log as u64) / 1_000_000) as u32
+	}
+}
+
+fn compute_points(input: &INposInput) -> Vec<(u32, u32)> {
```

Comment:
So I assume that if the arithmetic in such functions overflow (I know that a sensible curve should work fine, but assume the user makes a mistake) then it will be a compile-time error. Yet, we might have silent errors where `as` conversion is lossy (I haven't found a concrete example but I sense the potential --worth double-checking) is used. To also prevent such wrong curves to be compiled, I would recommend replacing a potentially wrong as with `try_into().unwrpa_or_else(|_| compile_error!(...))`

Reply:
done